Feature/client update after db update

kebyo opened the pull request:
1. Создал класс для того, чтобы хранить клиентов
2. Добавил отправку обновленного воркспейса клиенту

kebyo committed:
Create interface for Client

kebyo committed:
Class for clients

kebyo committed:
Addded method Add

kebyo committed:
Added route for adding workspace

kebyo committed:
Fixed JSDoc

kebyo committed:
Added clients saving

kebyo committed:
Update agent.ts

1. Added sending updated workspace to client
2. Added method for adding workspace

kebyo committed:
Disabled lint

kebyo committed:
Fixed dotenv options

neSpecc commented on src/index.ts:
```diff
@@ -15,12 +18,20 @@ const server = new ws.Server({
 });
 
-server.on('connection', (socket) => {
+const clients: Clients = new Clients();
```
не описано

neSpecc commented on src/types/client.ts:
```diff
@@ -0,0 +1,17 @@
+import ws from 'ws';
+
+/**
+ * Interface for client
```
непонятное описание

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -14,9 +16,35 @@ export default class AgentController {
     const workspace = await WorkspacesService.updateServices(req.headers.authorization, req.body.services);
 
+    if (workspace) {
+      const client: ws = clients.find(workspace.authToken);
```
тут не authToken должен быть, а интеграционный

kebyo replied:
messageId?

kebyo replied:
А как мне делать соответствие  между сокетом и интеграционным токеном? Парсить конфиг или как-то передавать в заголовке интеграционные токены при подключении клиента к серверу?

nikmel2803 replied:
Петя ответил ниже
Ещё тут есть проблема в том, что у тебя может быть несколько клиентов, получающих данные об одном воркспейсе, а тут у тебя используется find, которые возвращает лишь одного клиента
надо использовать filter и потом отправлять данные массиву клиентов

kebyo committed:
Created docs for routes

neSpecc commented on src/index.ts:
```diff
@@ -15,12 +18,20 @@ const server = new ws.Server({
 
+  const client: Client = {
+    socket,
+    authToken: req.headers.authorization!,
```
я думаю, где-то тут должна происходить проверка авторизации по заголовку и получение из базы (или конфига) воркспейса по authToken. Далее в Client передавать уже workspace._id

nikmel2803 commented on docs/routes.md:
```diff
@@ -0,0 +1,124 @@
+#Routes
```
```suggestion
# Routes
```

nikmel2803 commented on src/index.ts:
```diff
@@ -35,3 +46,5 @@ server.on('connection', (socket) => {
   });
 });
+
+export { clients };
```
Как-то не очень экспортировать тут что-то
1) можно сделать Clients синглтоном
2) можно инициализировать тут Clients и прокидывать этот объект либо res.locals или сделать поле context в Request

nikmel2803 commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,45 @@
+ * Clients class
+ */
+class Clients {
+  private clients: { [key: string]: ws};
```
тут не описано
ещё можно вместо `{ [key: string]: ws}` написать `Record<string, ws>`

nikmel2803 replied:
а ещё написать = {} и убрать конструктор тогда

nikmel2803 commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,45 @@
+   *
+   * @param authToken - Client authorization token
+   */
+  public find(authToken: string): ws {
```
как я написал выше, тут надо переделать, потому что по одному токену может быть несколько клиентов

kebyo committed:
Update docs/routes.md

Co-authored-by: Nikita Melnikov <nikmel2803@gmail.com>

kebyo committed:
Update client.ts

1. Updated JSDoc
2. Changed field

kebyo committed:
Used SingleTon

kebyo committed:
Merge branch 'feature/client-update-after-db-update' of https://github.com/codex-team/devops-toolbox-api into feature/client-update-after-db-update

neSpecc commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,51 @@
+ * Clients class
+ */
+class Clients {
+  private static singleTone: Clients;
```
не описаны свойства

neSpecc commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,51 @@
+import ws from 'ws';
+
+/**
+ * Clients class
```
непонятное описание

kebyo committed:
Update clients.ts

nikmel2803 commented on src/index.ts:
```diff
@@ -15,9 +20,29 @@ const server = new ws.Server({
+server.on('connection', async (socket: ws, req: express.Request) => {
   socket.send('Сonnected!');
 
+  const clients: Clients = Clients.getClients();
+
+  const authToken: string = req.headers.authorization!;
+
+  const workspaces: Workspace[] | null = await WorkspacesService.find({ authToken });
```
типы указывать тут не надо
они ни к чему, TS и так понимает всё

neSpecc replied:
не исправлено

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +20,29 @@ const server = new ws.Server({
+
+    clients.add(client);
+  } else {
+    console.log('Error authorization');
```
и соединение надо закрыть

kebyo committed:
Update index.ts

Added disconnection when client send wrong authorization token

neSpecc commented on docs/routes.md:
````diff
@@ -0,0 +1,124 @@
+    ```
+       
+       
+* **POST** - Добавляет воркспейс в БД
````
почему воркспейс добавляет `post /services`, а не `post /workspace` ?

kebyo replied:
У меня не работал скрипт Кости initDb.ts и я это добавил, чтобы тестить. Может это не нужно вообще?

neSpecc replied:
да, это пока не нужно. Давай удалим

neSpecc commented on docs/routes.md:
```diff
@@ -0,0 +1,124 @@
+                    "type": "nginx",
+                    "payload": [
+                        {
+                            "smth": "smth"
```
а структура payload где-то задокументирована?

kebyo replied:
Ну она зависит от типа сервиса. Где-то массив объектов, а где-то один объект. Я не стал более подробный запрос делать для примера, потому что там была какая-то инфа по серверу Кодекса. Подумал, что наверное не стоит это показывать.

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -14,9 +16,35 @@ export default class AgentController {
+    if (workspace) {
+      const clients: Client[] | undefined = Clients.getClients().find(workspace._id.toString());
+
+      clients?.forEach(client => client.socket.send(JSON.stringify(workspace)));
```
не понимаю эту строку. Тут просто в сокеты отправляется воркспейс. У нас же в сообщении должен быть messageId, type и payload. Как клиент поймет что это такое пришло?

kebyo replied:
А какой тогда будет messageId, если клиент не делает запрос? Это сообщение отправляется тогда, когда делается запрос от агента.

neSpecc replied:
`messageId` будет `null`. Так клиент поймет, что это не response, а сообщение, инициированное апи.  А `type` будет `workspace-update`. Так клиент поймет, как обработать payload.

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -14,9 +16,35 @@ export default class AgentController {
+
+    const workspaceDocument = await WorkspacesService.add(workspace);
+
+    res.json({
```
тоже не понятная структура объекта. Это что за интерфейс? Нужно указать тип

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +20,30 @@ const server = new ws.Server({
+server.on('connection', async (socket: ws, req: express.Request) => {
   socket.send('Сonnected!');
 
+  const clients: Clients = Clients.getClients();
```
не описаны константы

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +20,30 @@ const server = new ws.Server({
+    clients.add(client);
+  } else {
+    socket.send('Error authorization');
+    socket.close(1007);
```
magic number

neSpecc commented on src/routes/services.ts:
```diff
@@ -4,8 +4,13 @@ import AgentController from '../controllers/agent';
+/**
+ * POST request for add workspaces
+ */
+router.post('/', AgentController.addWorkspace);
```
не относится к сервисам

neSpecc commented on src/types/client.ts:
```diff
@@ -0,0 +1,17 @@
+   */
+  socket: ws;
+  /**
+   * Client workspaces
```
написано что это воркспейсы, а тип string[]

neSpecc commented on src/types/client.ts:
```diff
@@ -0,0 +1,17 @@
+  /**
+   * Client workspaces
+   */
+  workspaces: string[];
```
если это айдишники, то поле надо назвать workspaceIds

neSpecc commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * Field for a single instance of a class
+   */
+  private static singleTone: Clients;
```
обычно называют instance

kebyo committed:
Changed field name

neSpecc commented on src/utils/clients.ts:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Clients class for info about connected clients
+ */
+class Clients {
```
еще не очень нравится название сущности во множественном числе. Все таки класс представляет что-то одно. Может назвать ClientsList или ConnectedClientsList например или что-то типа того. А то путаница между Client и Clients.

neSpecc commented on src/index.ts:
```diff
@@ -34,4 +60,11 @@ server.on('connection', (socket) => {
+   * Сlient disconnects
+   */
+  socket.on('close', () => {
+    clients.remove(socket);
```
также нужно удалять клиента в обработчике onerror

kebyo committed:
Renamed Class

kebyo committed:
Remove client on error event

kebyo committed:
Moved route

kebyo committed:
Added route

kebyo committed:
Fixed JSDoc

kebyo committed:
Created response type

kebyo committed:
Renamed field

kebyo committed:
Updated JSDoc

kebyo committed:
Added method

kebyo committed:
Created docs for services

kebyo committed:
Update routes.md

kebyo committed:
Remove route /workspaces

kebyo committed:
Renamed interface

kebyo committed:
Fixed response

neSpecc commented on src/types/response.ts:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Interface for response
+ */
+interface Response {
```
не совсем корректное название. У `response` нету поля `type`. Это лучше назвать `Message`. А еще лучше сделать отдельные интерфейсы:

```ts
/**
 * Message can contain any data at payload
 */
export type MessagePayload = object;

/**
 * Any client-server message should fit this structure
 */
interface Message {
  payload: MessagePayload;
}

/**
 * Messages from the API sent in response to some client's message
 */
interface ResponseMessage extends Message {
   messageId: string; // <-- потом перенесем это в Message. По-хорошему, API тоже должен генерировать id сообщения, чтобы клиент мог на него ответить. 
}

/**
 * Message that initiated by the API
 */
interface OutgoingMessage extends Message {
  type: string;
}
```

neSpecc replied:
ну и `IncomingMessage` надо сделать еще

kebyo replied:
1. А почему API должен генерировать id сообщения? Мы разве не используем id сообщения, который нам пришел? Как клиент поймет, по другому id сообщения, что это конкретный "ответ" на его конкретный "запрос"?
2. Зачем клиенту отвечать на сообщения от API?

neSpecc replied:
1. Это только в случае `OutgoingMessage`, когда сообщение инициирует сам бэкенд
2. Например, в следующих версиях протокола может появиться подтверждение получения сообщений. То есть когда клиент получает сообщение, он должен отправить в ответ `ack` с `id` этого сообщения. В противном случае, бэкенд будет слать это сообщение повторно, пока не получит `ack`. Так, например, работает API Telegram. Это нужно для гарантии доставки сообщений, даже при проблемах с интернетом.

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -12,7 +16,21 @@ export default class AgentController {
+    if (workspace) {
+      const clients: Client[] | undefined = ClientsList.getClients().find(workspace._id.toString());
+
+      const wsResponse: Response = {
```
тут все еще не идеально: формирование сообщений должно быть где-то в одном месте. В классе, отвечающем за транспортный уровень. А сейчас у нас они формируются, по сути, хардкодом в разных местах (`controllers/agent.ts`, `controllers/workspace.ts`, `src/index.ts`).

Подумай, куда можно это вынести

neSpecc commented on src/controllers/workspaces.ts:
```diff
@@ -11,4 +12,24 @@ export default class WorkspacesController {
+
+    const workspaceDocument: Workspace | null = await WorkspacesService.add(workspace);
+
+    res.json({
```
вот тут тоже ответ не соответствует протоколу

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +19,43 @@ const server = new ws.Server({
   socket.send('Сonnected!');
 
+  /**
+   * Current connected clients
```
```suggestion
   * Currently connected clients
```

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +19,43 @@ const server = new ws.Server({
+  const clients = ClientsList.getClients();
+
+  /**
+   * Connecting client authorization token
```
```suggestion
   * Connected client's authorization token
```

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +19,43 @@ const server = new ws.Server({
+  const authToken = req.headers.authorization!;
+
+  /**
+   * Connecting client workspaces
```
```suggestion
   * Connected client's workspaces list
```

neSpecc commented on src/index.ts:
```diff
@@ -34,4 +72,18 @@ server.on('connection', (socket) => {
+  });
+
+  /**
+   * Error
```
```suggestion
   * Sockets error handler 
```

neSpecc commented on src/index.ts:
```diff
@@ -34,4 +72,18 @@ server.on('connection', (socket) => {
   });
+
+  /**
+   * Сlient disconnects
```
```suggestion
   * Сlient disconnecting handler
```

neSpecc commented on src/index.ts:
```diff
@@ -15,9 +19,43 @@ const server = new ws.Server({
+
+    clients.add(client);
+  } else {
+    socket.send('Error authorization');
```
ответ не соответствует нашему протоколу. Думаю, он тут вообще не нужен, кода достаточно

kebyo committed:
Update src/index.ts

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

kebyo committed:
Update src/index.ts

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

kebyo committed:
Update src/index.ts

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

kebyo committed:
Update src/index.ts

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

kebyo committed:
Update src/index.ts

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

kebyo committed:
Create incomingMessage.ts

kebyo committed:
Create message.ts

kebyo committed:
Create outgoingMessage.ts

kebyo committed:
Create requestMessage.ts

kebyo committed:
Create responseMessage.ts

kebyo committed:
Delete old types

kebyo committed:
Added easy import

kebyo committed:
Fixed method name

kebyo committed:
Created class Server

kebyo committed:
Merge branch 'feature/client-update-after-db-update' of https://github.com/codex-team/devops-toolbox-api into feature/client-update-after-db-update

kebyo committed:
Fixed JSDoc

neSpecc commented on src/utils/server.ts:
```diff
@@ -0,0 +1,135 @@
+/**
+ * Class Server (Transport level)
+ */
+export default class Server {
```
Почему сервер? Давай назовем этот класс Transport. 

и я бы положил его (и типы, связанные с ним) в папку `/utils/protocol/`. В будущем его можно будет вынести в отдельный npm-пакет и использовать в других проектах

neSpecc commented on src/utils/server.ts:
```diff
@@ -0,0 +1,135 @@
+   * @param socket - socket
+   * @param req - additional GET request from client
+   */
+  public static async connection(socket: ws, req: express.Request): Promise<void> {
```
почему все через статик?

neSpecc commented on src/types/outgoingMessage.ts:
```diff
@@ -0,0 +1,10 @@
+ */
+import { Message } from './message';
+
+interface OutgoingMessage extends Message {
```
не описано

neSpecc commented on src/types/message.ts:
```diff
@@ -0,0 +1,15 @@
+ * Any client-server message should fit this structure
+ */
+export interface Message {
+  payload: MessagePayload;
```
свойства надо описать

neSpecc commented on src/types/message.ts:
```diff
@@ -0,0 +1,15 @@
+ * Message can contain any data at payload
+ */
+export type MessagePayload = {
+  workspaces?: Workspace[] | null;
```
это хардкод. По сути это типы либы Transport, тут не должно быть конкретных описаний пэйлоадов. Они должны быть где-то выше

neSpecc commented on src/types/incomingMessage.ts:
```diff
@@ -0,0 +1,10 @@
+ */
+import RequestMessage from './requestMessage';
+
+interface IncomingMessage extends RequestMessage {
```
не описан класс и свойства

neSpecc commented on src/index.ts:
```diff
@@ -15,23 +14,7 @@ const server = new ws.Server({
+/**
+ *  Client connects
+ */
+server.on('connection', Server.connection);
```
это (и `new ws.Server...`) тоже должно быть убрано внутрь транспорта.

В точке входа должно остаться что-то типа:

```ts
this.transport = new Transport({ 
  endpoint: ''
});
```

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -12,7 +15,13 @@ export default class AgentController {
+    if (workspace) {
+      const clients: Client[] | undefined = ClientsList.getAll().find(workspace._id.toString());
+
+      clients?.forEach(client => Server.send(client.socket, null, { workspace }));
```
а где type? у каждого сообщения должен быть тип. И его формирование — это не задача Транспорта, в отличие от `messageId`. Его нужно взять из какой-то константы, которая лежит вне транспорта (тк это логика нашего приложения, а не либы Transport)

neSpecc commented on src/utils/server.ts:
```diff
@@ -0,0 +1,135 @@
+   * @param messageId - message Id
+   * @param payload - payload
+   */
+  public static send(socket: ws, messageId: string | null, payload: object): void {
```
`messageId` — это внутренняя фича протокола, то есть она должна формироваться внутри, а не передаваться снаружи.

neSpecc commented on src/utils/server.ts:
```diff
@@ -0,0 +1,135 @@
+    /**
+     * Connected client's workspaces list
+     */
+    const workspaces = await WorkspacesService.find({ authToken });
```
протокол ничего не должен знать о воркспейсах и другой логике приложения

neSpecc commented on src/utils/server.ts:
```diff
@@ -0,0 +1,135 @@
+
+    const payload: MessagePayload = {};
+
+    switch (dataObj.type) {
```
это явно не задача протокола (утилиты).

kebyo committed:
Updated JSDoc

kebyo committed:
Added new field

kebyo committed:
Added transport options type

kebyo committed:
Updated class Transport (old Server)

1. Renamed from Server to Transport
2. Fixed methods
3. Moved to protocol directory

kebyo committed:
Fixed filed

kebyo committed:
Moved protocol's type to src/protocol/types

kebyo committed:
Updated JSDoc

kebyo committed:
Create messageCreator.ts

kebyo committed:
Updated method send

kebyo committed:
Update agent.ts

neSpecc commented on src/utils/protocol/types/responseMessage.ts:
```diff
@@ -0,0 +1,13 @@
+  /**
+   * Message id, which we use to send response with this id (If API send message then messageId is null)
+   */
+  messageId: string | null;
```
тут не может быть null

neSpecc commented on src/utils/protocol/types/responseMessage.ts:
```diff
@@ -0,0 +1,13 @@
+ */
+interface ResponseMessage extends Message {
+  /**
+   * Message id, which we use to send response with this id (If API send message then messageId is null)
```
> (If API send message then messageId is null)

Это уже не ResponseMessage

neSpecc commented on src/utils/protocol/messageCreator.ts:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Class for creating message
+ */
+export default class MessageCreator {
```
давай избавимся от этого класса. Контроллеры не должны знать, какой тип сообщения они отправляют — то есть параметра `messageId` не должно быть. Формирование id сообщения это внутренняя логика транспорта. 

То есть лучше сделать в самом Transport.ts два отдельных метода: `send(type, payload)` (отправляет OutgoingMessage) и `respond(messageId, payload)`  (отравляет ResponseMessage)

neSpecc commented on src/utils/protocol/transport.ts:
```diff
@@ -0,0 +1,137 @@
+   * @param socket - client
+   * @param message - message
+   */
+  public static send(socket: ws, message: ResponseMessage | OutgoingMessage): void {
```
смущает что статик. Этот метод зависит от текущего инстанса, тк зависит от опций, переданных в конструктор. А раз зависит от инстанса, значит это не статик.

kebyo replied:
Он статик, потому что в методе onmessage один из аргументов сокет this и я не знаю как вызвать по другому метод send, если внутри метода onmessage this это сокет.

neSpecc commented on src/controllers/agent.ts:
```diff
@@ -12,7 +16,18 @@ export default class AgentController {
+    if (workspace) {
+      const clients: Client[] | undefined = ClientsList.getAll().find(workspace._id.toString());
+
+      clients?.forEach(client => Transport.send(client.socket, MessageCreator.create(
```
такое ощущение, что Client — это тоже часть протокола. А значит передавать в send лучше `client` прямо, он знает что в нем есть `.socket`

kebyo committed:
Merge branch 'main' into feature/client-update-after-db-update

kebyo committed:
Update JSDoc

kebyo committed:
Fixed MessagePayload

kebyo committed:
Move protocol types and class to /protocol/types

kebyo committed:
Delete messageCreator.ts

kebyo committed:
Update transport.ts

kebyo committed:
Use fixed Transport method

kebyo committed:
Add authorization

neSpecc commented on src/utils/protocol/transport.ts:
```diff
@@ -0,0 +1,168 @@
+      console.log(`⚡️[server]: Server is running at ws://localhost:${options.port}/${options.path}`);
+    });
+
+    /**
```
wtf?
